lib/Fmt_ast.ml

```diff
@@ -2893,6 +2894,36 @@ and fmt_expression c ?(box = true) ?(pro = noop) ?eol ?parens
   | Pexp_indexop_access x ->
       pro $ fmt_indexop_access c ctx ~fmt_atrs ~has_attr ~parens x
   | Pexp_hole -> pro $ hvbox 0 (fmt_hole () $ fmt_atrs)
+  | Pexp_beginend ({pexp_desc= Pexp_match (e0, cs); _} as e) ->
+      let xexp = sub_exp ~ctx e in
+      let ctx = Exp e in
+      hvbox 0
+        ( fmt_match c ~pro ~parens:false ctx xexp cs e0
+            (hvbox 0
+               ( str "begin"
+               $ fmt_extension_suffix c ext
+               $ fmt_atrs $ break 1 0 $ str "match" ) )
+        $ break 1 0 $ str "end" )
+  | Pexp_beginend
+      { pexp_desc=
+          Pexp_extension
+            ( ext_inner
+            , PStr
+                [ ( { pstr_desc=
+                        Pstr_eval
+                          (({pexp_desc= Pexp_match (e0, cs); _} as e1), _)
+                    ; pstr_loc= _ } as stru ) ] )
+      ; _ }
+    when Source.extension_using_sugar ~name:ext_inner ~payload:e1.pexp_loc ->
+      let ctx = Str stru in
+      let xexp = sub_exp ~ctx e1 in
```

Comment:
I'd really want to avoid the code duplication. This should be moved to a new function.